File: LinkedList/1)SinglyLinkedList/Sll.js
// Node class
class Node {
    constructor(data) {
        this.data = data; // Node's data
        this.next = null; // Pointer to the next node
    }
}

// LinkedList class
class LinkedList {
    constructor() {
        this.head = null; // Head of the list
    }

    // Insert a node at the beginning
    insertAtBeginning(data) {
        const newNode = new Node(data); // Create a new node
        newNode.next = this.head; // Link the new node to the current head
        this.head = newNode; // Update the head to the new node
    }

    // Insert a node at the end
    insertAtEnd(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode; // If the list is empty, make the new node the head
            return;
        }
        let current = this.head;
        while (current.next) {
            current = current.next; // Traverse to the last node
        }
        current.next = newNode; // Link the last node to the new node
    }

    // Method to insert at a specific position iteratively
    insertAtPosition(data, position) {
        // Create the new node to be inserted
        const newNode = new Node(data);

        // If inserting at position 0, make the new node the head
        if (position === 0) {
            newNode.next = this.head;
            this.head = newNode;
            return;
        }

        let current = this.head;
        let currentPosition = 0;

        // Traverse the list to find the position before where we want to insert
        while (current !== null && currentPosition < position - 1) {
            current = current.next;
            currentPosition++;
        }

        // If the position is valid (within the bounds of the list)
        if (current !== null) {
            // Insert the new node at the desired position
            newNode.next = current.next;
            current.next = newNode;
        } else {
            console.log("Position out of bounds");
        }
    }

    // Display the linked list
    display() {
        let current = this.head;
        let result = "";
        while (current) {
            result += current.data + " -> ";
            current = current.next;
        }
        console.log(result + "null");
    }
}

// Example usage
const list = new LinkedList();

// Add at the beginning
list.addAtBeginning(30);
list.addAtBeginning(20);
list.addAtBeginning(10);
console.log("After adding at the beginning:");
list.display();

// Add at the end
list.addAtEnd(40);
list.addAtEnd(50);
console.log("After adding at the end:");
list.display();

File: LinkedList/1)SinglyLinkedList/Sll.java
class Node {
    int data;
    Node next;

    // Constructor
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head; // Head of the list

    public LinkedList() {
        this.head = null;
    }

    // Method to insert a new node at the beginning
    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = this.head; // Point the new node to the current head
        this.head = newNode; // Update head to the new node
    }

    // Method to insert a new node at the end
    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // Method to insert at a specific position iteratively
    public void insertAtPosition(int data, int position) {
        // Create the new node to be inserted
        Node newNode = new Node(data);

        // If inserting at position 0, make the new node the head
        if (position == 0) {
            newNode.next = head;
            head = newNode;
            return;
        }

        Node current = head;
        int currentPosition = 0;

        // Traverse the list to find the position before where we want to insert
        while (current != null && currentPosition < position - 1) {
            current = current.next;
            currentPosition++;
        }

        // If the position is valid (within the bounds of the list)
        if (current != null) {
            // Insert the new node at the desired position
            newNode.next = current.next;
            current.next = newNode;
        } else {
            System.out.println("Position out of bounds");
        }
    }

    // Method to display the linked list
    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    // Method to delete a node with a specific value
    public void delete(int value) {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        if (head.data == value) {
            head = head.next;
            return;
        }
        Node current = head;
        while (current.next != null && current.next.data != value) {
            current = current.next;
        }
        if (current.next == null) {
            System.out.println("Value not found.");
        } else {
            current.next = current.next.next;
        }
    }
}

class Main {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.insertAtEnd(10);
        list.insertAtEnd(20);
        list.insertAtEnd(30);

        System.out.println("Original List:");
        list.display();

        list.delete(20);
        System.out.println("After Deletion:");
        list.display();
    }
}
